Added $resource parameter to getBaseSelfUrl method

boris-glumpler opened the pull request:
Duh! Obviously didn't think this all the way through in #4. Anyways, the $resource parameter now gets passed as well to getBaseSelfUrl.

boris-glumpler committed:
Added $resource parameter to getBaseSelfUrl method

boris-glumpler commented:
Cheers!

Sent by iPhone.

> On 10 May 2015, at 20:58, neomerx notifications@github.com wrote:
> 
> Merged #5.
> 
> —
> Reply to this email directly or view it on GitHub.

neomerx commented:
@shabushabu I've added error encoding to encoder and renamed the very last parameter for `encode` method from  `EncodingOptions` to `EncodingParameters`. If you haven't started using 'sparse' and 'fields sets' features it shouldn't affect you.
Version was bumped to 0.2. Please update.

boris-glumpler commented:
Will do, thanks! I've started the API using Fractal, so still in the process of turning transformers into schemas. Will still be a while until I get to sparse and such. 

Sent by iPhone.

> On 10 May 2015, at 21:20, neomerx notifications@github.com wrote:
> 
> @shabushabu I've added error encoding to encoder and renamed the very last parameter for encode method from EncodingOptions to EncodingParameters. If you haven't started using 'sparse' and 'fields sets' features it shouldn't affect you.
> Version was bumped to 0.2. Please update.
> 
> —
> Reply to this email directly or view it on GitHub.

neomerx commented:
@shabushabu I'm working on adding request/response specifics from the spec and my main framework are Laravel5 and Lumen. So if anything good comes out it might be beneficial to you.

boris-glumpler commented:
Sounds good! I'm happy to help as well where I can.
Is the request/response stuff gonna be coupled to Eloquent at all? Cause I'm using Doctrine on this project ;)

Sent by iPhone.

> On 10 May 2015, at 21:35, neomerx notifications@github.com wrote:
> 
> @shabushabu I'm working on adding request/response specifics from the spec and my main framework is Laravel5 and Lumen. So if anything good comes out it might be beneficial to you.
> 
> —
> Reply to this email directly or view it on GitHub.

neomerx commented:
@shabushabu No, it won't be coupled with any frameworks. The goal is to move to full and feature complete implementation as close as possible but
- do not reinvent the wheel (e.g. writing one more symfony/laravel/whatever)
- stay framework agnostic
- make it as easy as possible to integrate to symfony/laravel/whatever

boris-glumpler commented:
That all sounds good!

boris-glumpler commented:
Just checked the source and you've removed $resource from `getBaseSelfUrl()`, but it's still being passed into it in `getSelfUrl()`. This is the only place `getBaseSelfUrl()` is being used, so why not just pass it in, like you do with `getId()`?

So, how would I go about getting the resource object from the input parameters? Isn't immediately obvious to me.

neomerx commented:
Scrutinizer gave an issue about unused parameter. However when I removed it (but kept in invoke) I got another issue from scrutinizer. So it didn't worked out. I'm OK to get it back if necessary.
You just override the method and declare the param. Like so

``` php
protected function getBaseSelfUrl($resource) {
   // $resource will be not null. just try it
}
```

neomerx commented:
Please let me know if you have any issues

boris-glumpler commented:
Yeah, I figured that out, thanks!. I should know better than to write code after 10 pm.

neomerx commented:
@shabushabu I've done some pretty nice job in `web` branch and I think it might be helpful for you. Actually it can save you a ton of time and make adding services **extremely** easy :wink: 

For example minimal controller to handle `GET resource` requests might be as simple as this

``` php
class AuthorsController extends Controller
{
    public function show($id)
    {
        return $this->getResponse(Author::findOrFail($id));
    }
}
```

It's not fully documented (yet!)

I've forked default Laravel app and added a few JSON API services
- CRUD operations supported
- nice exception handling
- JSON API Errors supported

it's [here](https://github.com/neomerx/limoncello-collins/tree/limoncello)

Please note it's in `limoncello` branch. You can see [my commits](https://github.com/neomerx/limoncello-collins/commits/limoncello) which are actually 7-step 'how to create JSON API app with Laravel'
- Add settings and framework integration https://github.com/neomerx/limoncello-collins/commit/65016569da6680fdf93b2c47eddf17f8759d2e5d
- Add models, migration and seeds https://github.com/neomerx/limoncello-collins/commit/b02f5a091ee561a6354c5adb7dc36156a12b5f0a
- Add model schemas https://github.com/neomerx/limoncello-collins/commit/00cb697057c5d10089c4345ebc22bd97963cdd19
- Add exception handling https://github.com/neomerx/limoncello-collins/commit/c31866459feb3db3a99cf48cc61823ff57fec87d
- Add routes and controllers https://github.com/neomerx/limoncello-collins/commit/951e312c6fd904e0bb1a2b4f63c00479d16b9264
- Update composer.json (for now it uses dev versions) https://github.com/neomerx/limoncello-collins/commit/165f9f0c072cf22764eead963d50fa4db3f0c56c
- Add tests https://github.com/neomerx/limoncello-collins/commit/661655d9e2b12f53a197dde21b0a4461bed169d0

You can run this example

```
$ git clone https://github.com/neomerx/limoncello-collins && cd limoncello-collins/ && git checkout limoncello
$ touch storage/database.sqlite 
$ php artisan migrate --seed
$ php artisan serve --port=8888
```

And request API with `curl`

```
$ curl -X GET -H "Content-Type: application/vnd.api+json" -H "Accept: application/vnd.api+json" http://localhost:8888/authors/1
```

which should return

``` json
{
    "data": {
        "type": "authors",
        "id": "1",
        "attributes": {
            "first_name": "Dan",
            "last_name": "Gebhardt"
        },
        "links": {
            "self": "http:\/\/localhost:8888\/authors\/1",
            "posts": {
                "linkage": {
                    "type": "posts",
                    "id": "1"
                }
            }
        }
    }
}
```

boris-glumpler commented:
Sounds great! Will check it out as soon as I get the chance!

Sent by iPhone.

> On 14 May 2015, at 04:20, neomerx notifications@github.com wrote:
> 
> @shabushabu I've done some pretty nice job in web branch and I think it might be helpful for you. Actually it can save you a ton of time and make adding services extremely easy 
> 
> For example minimal controller to handle GET resource requests might be as simple as this
> 
> class AuthorsController extends Controller
> {
>     public function show($id)
>     {
>         return $this->getResponse(Author::findOrFail($id));
>     }
> }
> It's not fully documented (yet!)
> 
> I've forked default Laravel app and added a few JSON API services
> 
> CRUD operations supported
> nice exception handling
> JSON API Errors supported
> it's here
> 
> Please note it's in limoncello branch. You can see my commits which are actually 7-step 'how to create JSON API app with Laravel'
> 
> Add settings and framework integration neomerx/limoncello-collins@6501656
> Add models, migration and seeds neomerx/limoncello-collins@b02f5a0
> Add model schemas neomerx/limoncello-collins@00cb697
> Add exception handling neomerx/limoncello-collins@c318664
> Add routes and controllers neomerx/limoncello-collins@951e312
> Update composer.json (for now it uses dev versions) neomerx/limoncello-collins@165f9f0
> Add tests neomerx/limoncello-collins@661655d
> You can run this example
> 
> $ composer create-project neomerx/limoncello-collins
> $ touch storage/database.sqlite 
> $ php artisan migrate --seed
> $ php artisan serve --port=8888
> And request API with curl
> 
> $ curl -X GET -H "Content-Type: application/vnd.api+json" -H "Accept: application/vnd.api+json" http://localhost:8888/authors/1
> which should return
> 
> {
>     "data": {
>         "type": "authors",
>         "id": "1",
>         "attributes": {
>             "first_name": "Dan",
>             "last_name": "Gebhardt"
>         },
>         "links": {
>             "self": "http:\/\/localhost:8888\/authors\/1",
>             "posts": {
>                 "linkage": {
>                     "type": "posts",
>                     "id": "1"
>                 }
>             }
>         }
>     }
> }
> —
> Reply to this email directly or view it on GitHub.

boris-glumpler commented:
Looked it quickly over and it looks really great. Should help me a lot. Need to sort out some visa stuff in the next few days, but will get back to it shortly. I'll let you know how I get on.
